File: backend.scala/app/com/skewwhiffy/auraltester/clefs/Clef.scala
package com.skewwhiffy.auraltester.clefs

import com.skewwhiffy.auraltester.notes.{AbsoluteNote, Interval, Note}
import com.skewwhiffy.auraltester.scales.NoteSequence

case class Clef(
  abc: String,
  lowLedgerNote: AbsoluteNote,
  highLedgerNote: AbsoluteNote
) {
  def displayName: String = abc

  def notes: List[NoteSequence] = List(
    NoteSequence(AbsoluteNote.middleC.withLyric("Middle~C")),
    lineNotes,
    spaceNotes,
    lowerLedgerNotes,
    upperLedgerNotes,
  )

  def lineNotes: NoteSequence = ClefLineNotes(this)

  def spaceNotes: NoteSequence = ClefSpaceNotes(this)

  def lowerLedgerNotes: NoteSequence = ClefLowerLedgerNotes(this)

  def upperLedgerNotes: NoteSequence = ClefUpperLedgerNotes(this)

  def getNoteNearBottom(note: Note): AbsoluteNote = {
    val candidateStartingNote = AbsoluteNote(note, lowLedgerNote.octave)
    if (candidateStartingNote < lowLedgerNote) candidateStartingNote + Interval.perfect(8) else candidateStartingNote
  }
}

File: backend/src/main/kotlin/com/skewwhiffy/notation/model/Clef.kt
package com.skewwhiffy.notation.model

import com.skewwhiffy.notation.model.abc.ClefLineNotes
import com.skewwhiffy.notation.model.abc.ClefLowerLedgerNotes
import com.skewwhiffy.notation.model.abc.ClefSpaceNotes
import com.skewwhiffy.notation.model.abc.ClefUpperLedgerNotes

data class Clef(
  val abc: String,
  val lowLedgerNote: AbsoluteNote,
  val highLedgerNote: AbsoluteNote,
) {
  val displayName: String = abc

  private val lineNotes: NoteSequence = ClefLineNotes(this)

  private val spaceNotes: NoteSequence = ClefSpaceNotes(this)

  private val lowerLedgerNotes: NoteSequence = ClefLowerLedgerNotes(this)

  private val upperLedgerNotes: NoteSequence = ClefUpperLedgerNotes(this)


  val notes: List<NoteSequence> = listOf(
    NoteSequence.of(AbsoluteNote.middleC.withLyric("Middle~C")),
    lineNotes,
    spaceNotes,
    lowerLedgerNotes,
    upperLedgerNotes,
  )

  fun getNoteNearBottom(note: Note): AbsoluteNote {
    val candidateStartingNote = AbsoluteNote(note, lowLedgerNote.octave)
    return if (candidateStartingNote < lowLedgerNote) candidateStartingNote + Interval.perfect(8) else candidateStartingNote
  }
}